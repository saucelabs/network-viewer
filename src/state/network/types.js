export const UPDATE_DATA = 'UPDATE_DATA';
export const UPDATE_SEARCH = 'UPDATE_SEARCH';
export const UPDATE_SORT = 'UPDATE_SORT';
export const UPDATE_FILTER = 'UPDATE_FILTER';
