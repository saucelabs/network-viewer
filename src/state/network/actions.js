import {
  UPDATE_DATA, UPDATE_SEARCH, UPDATE_SORT, UPDATE_FILTER,
} from './types';

export const updateData = dispatch => payload => dispatch({
  type: UPDATE_DATA,
  payload,
});

export const updateSearch = dispatch => payload => dispatch({
  type: UPDATE_SEARCH,
  payload,
});

export const updateSort = dispatch => payload => dispatch({
  type: UPDATE_SORT,
  payload,
});

export const updateFilter = dispatch => payload => dispatch({
  type: UPDATE_FILTER,
  payload,
});
