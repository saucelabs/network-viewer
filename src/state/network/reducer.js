import { Map, List } from 'immutable';

import { filterData, sortBy, prepareViewerData, calculateTimings, getSummary } from './../../utils';
import * as types from './types';

const initialState = new Map({
  data: new List(),
  actualData: new List(),
  totalNetworkTime: null,
  dataSummary: new Map(),
  sort: {
    key: 'startedDateTime',
    isAcs: true,
  },
  search: {
    name: 'url',
    value: '',
  },
  filter: {
    name: null,
    value: null,
  },
  errorFilter: false,
  error: null,
  loading: false,
  scrollToIndex: null,
  selectedReqIndex: null,
  showReqDetail: false,
  reqDetail: null,
});

const reducer = (state = initialState, { type, payload }) => {
  switch (type) {
    case types.UPDATE_DATA: {
      return state.withMutations((newState) => {
        const sort = state.get('sort');
        const {
          data,
          totalNetworkTime,
          totalRequests,
          totalTransferredSize,
          totalUncompressedSize,
          finishTime,
        } = prepareViewerData(payload.entries);
        const sortedData = new List(sortBy(data, sort.key, sort.isAcs));
        newState
          .set('error', null)
          .set('data', sortedData)
          .set('actualData', sortedData)
          .set('totalNetworkTime', totalNetworkTime)
          .set('dataSummary', new Map({
            totalRequests,
            totalTransferredSize,
            totalUncompressedSize,
            finishTime,
            timings: calculateTimings(payload.pages),
            finish: finishTime,
          }));
      });
    }
    case types.UPDATE_SEARCH: {
      return state.withMutations((newState) => {
        const data = filterData({
          data: state.get('actualData'),
          filter: state.get('filter'),
          search: payload,
          errorFilter: state.get('errorFilter'),
        });
        const summary = getSummary(data);
        newState
          .set('search', payload)
          .set('data', data)
          .setIn(['dataSummary', 'totalRequests'], summary.totalRequests)
          .setIn(['dataSummary', 'totalTransferredSize'], summary.totalTransferredSize)
          .setIn(['dataSummary', 'totalUncompressedSize'], summary.totalUncompressedSize);
      });
    }
    case types.UPDATE_FILTER: {
      return state.withMutations((newState) => {
        const data = filterData({
          data: state.get('actualData'),
          filter: payload,
          search: state.get('search'),
          errorFilter: state.get('errorFilter'),
        });
        const summary = getSummary(data);
        newState
          .set('filter', payload)
          .set('data', data)
          .setIn(['dataSummary', 'totalRequests'], summary.totalRequests)
          .setIn(['dataSummary', 'totalTransferredSize'], summary.totalTransferredSize)
          .setIn(['dataSummary', 'totalUncompressedSize'], summary.totalUncompressedSize);
      });
    }
    case types.UPDATE_ERROR_FILTER: {
      return state.withMutations((newState) => {
        const data = filterData({
          data: state.get('actualData'),
          filter: state.get('filter'),
          search: state.get('search'),
          errorFilter: payload,
        });
        const summary = getSummary(data);
        newState
          .set('errorFilter', payload)
          .set('data', data)
          .setIn(['dataSummary', 'totalRequests'], summary.totalRequests)
          .setIn(['dataSummary', 'totalTransferredSize'], summary.totalTransferredSize)
          .setIn(['dataSummary', 'totalUncompressedSize'], summary.totalUncompressedSize);
      });
    }
    case types.UPDATE_SORT: {
      return state.withMutations((newState) => {
        newState
          .set('sort', payload)
          .set('data', sortBy(state.get('data'), payload.key, payload.isAcs));
      });
    }
    case types.FETCH_FILE.REQUEST: {
      return state.withMutations((newState) => {
        newState
          .set('error', null)
          .set('loading', true);
      });
    }
    case types.FETCH_FILE.SUCCESS: {
      return state.withMutations((newState) => {
        newState
          .set('error', null)
          .set('loading', false);
      });
    }
    case types.FETCH_FILE.FAILURE:
    case types.UPDATE_ERROR_MESSAGE: {
      return state.withMutations((newState) => {
        newState
          .set('error', payload)
          .set('loading', false);
      });
    }
    case types.UPDATE_SCROLL_TO_INDEX: {
      return state.withMutations((newState) => {
        newState
          .set('selectedReqIndex', payload ? payload.index : null)
          .set('reqDetail', payload);
      });
    }
    case types.SELECT_REQUEST: {
      return state.withMutations((newState) => {
        newState
          .set('selectedReqIndex', payload ? payload.index : null)
          .set('reqDetail', payload)
          .set('showReqDetail', !!payload);
      });
    }
    case types.RESET: {
      return initialState;
    }
    default:
      return state;
  }
};

export { initialState, reducer };
export default reducer;
