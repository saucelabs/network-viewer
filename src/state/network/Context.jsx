import React, { useContext } from 'react';

import { actionsWrapper } from './../../utils';
import * as actions from './actions';

export const NetworkContext = React.createContext();

export const useNetwork = () => {
  const context = useContext(NetworkContext);
  if (!context) {
    throw new Error('useNetwork must be used within a NetworkProvider');
  }
  const [state, dispatch] = context;

  const wrappedActions = actionsWrapper({
    updateData: actions.updateData,
    updateSearch: actions.updateSearch,
    updateSort: actions.updateSort,
    updateFilter: actions.updateFilter,
    updateErrorMessage: actions.updateErrorMessage,
    selectRequest: actions.selectRequest,
    updateScrollToIndex: actions.updateScrollToIndex,
    updateErrorFilter: actions.updateErrorFilter,
    resetState: actions.resetState,
  })(dispatch, state);

  return {
    state,
    dispatch,
    actions: wrappedActions,
  };
};
