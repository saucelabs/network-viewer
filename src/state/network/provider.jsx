import React, { useReducer, useContext, useMemo } from 'react';

import { reducer, initialState } from './reducer';
import * as actions from './actions';
import { actionsWrapper } from '../../utils';

export const NetworkContext = React.createContext();

export const NetworkProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const value = useMemo(() => [state, dispatch], [state]);
  return <NetworkContext.Provider value={value} {...props} />;
};

export const useNetwork = () => {
  const context = useContext(NetworkContext);
  if (!context) {
    throw new Error('useNetwork must be used within a NetworkProvider');
  }
  const [state, dispatch] = context;

  return {
    state,
    dispatch,
    actions: actionsWrapper({
      updateData: actions.updateData,
      updateSearch: actions.updateSearch,
      updateSort: actions.updateSort,
      updateFilter: actions.updateFilter,
      errorNotification: actions.errorNotification,
      warningNotification: actions.warningNotification,
      successNotification: actions.successNotification,
      infoNotification: actions.infoNotification,
      dismissNotification: actions.dismissNotification,
    })(dispatch, state),
  };
};
