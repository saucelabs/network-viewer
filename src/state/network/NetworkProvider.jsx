import React, { useEffect, useReducer, useMemo } from 'react';
import PropTypes from 'prop-types';

import { reducer, initialState as defaultState } from './reducer';
import { updateData, fetchFile, updateScrollToIndex } from './actions';
import { NetworkContext } from './Context';
import { findRequestIndex } from '../../utils';
import { ROW_ID_PREFIX } from '../../constants';

const NetworkProvider = (props) => {
  const {
    data,
    file,
    fetchOptions,
    initialState,
    scrollTimeStamp,
    scrollRequestPosition,
    autoHighlightChange,
    onDataLoaded,
    onDataError,
    onPause,
    onResume,
    onReset,
  } = props;

  const [state, dispatch] = useReducer(reducer, initialState);
  const callbacks = {
    onPause,
    onResume,
    onReset,
  };
  const value = useMemo(() => [state, dispatch, callbacks], [state]);
  const selectedReqIndex = value[0].get('selectedReqIndex');
  const requestData = value[0].get('data');
  const showReqDetail = value[0].get('showReqDetail');
  const actualData = value[0].get('actualData');
  const error = value[0].get('error');

  // Update data onChange of network data
  useEffect(() => {
    if (data && data.log && data.log.entries) {
      updateData(dispatch)(data);
    }
  }, [data]);

  // Fetch HAR file onChange of file prop
  useEffect(() => {
    if (file) {
      fetchFile(dispatch)(file, fetchOptions);
    }
  }, [file]);

  useEffect(() => {
    if (actualData.size && onDataLoaded) {
      onDataLoaded(actualData);
    }
  }, [actualData]);

  useEffect(() => {
    if (error && onDataError) {
      onDataError(error);
    }
  }, [error]);

  // Find nearby request-rowId and update scrollIndex on scrollTimeStamp receive
  useEffect(() => {
    const shouldChangeHighlight = showReqDetail ? autoHighlightChange : true;
    if (scrollTimeStamp && shouldChangeHighlight) {
      const reqIndex = findRequestIndex({
        data: requestData,
        timestamp: scrollTimeStamp,
        position: scrollRequestPosition,
      });
      if (reqIndex || reqIndex === 0) {
        updateScrollToIndex(dispatch)(requestData.get(reqIndex));
      }
    }
  }, [scrollTimeStamp, actualData]);

  // Scroll to request row onChange of scrollToIndex
  // setTimeout is required when reqDetail is visible,
  // it allows DOM to adjust before we scroll it to highlighted request
  useEffect(() => {
    if (selectedReqIndex) {
      setTimeout(() => {
        const element = document.getElementById(ROW_ID_PREFIX + selectedReqIndex);
        if (element) {
          element.scrollIntoView({
            alignToTop: true,
            behavior: 'smooth',
          });
        }
      }, 300);
    }
  }, [selectedReqIndex]);

  return (
    <NetworkContext.Provider
      value={value}
      {...props}
    />
  );
};

NetworkProvider.propTypes = {
  autoHighlightChange: PropTypes.bool,
  data: PropTypes.object,
  fetchOptions: PropTypes.object,
  file: PropTypes.string,
  initialState: PropTypes.object,
  onDataError: PropTypes.func,
  onDataLoaded: PropTypes.func,
  onPause: PropTypes.func,
  onReset: PropTypes.func,
  onResume: PropTypes.func,
  scrollRequestPosition: PropTypes.oneOf(['before', 'after', 'near']),
  scrollTimeStamp: PropTypes.number,
};

NetworkProvider.defaultProps = {
  autoHighlightChange: false,
  data: null,
  fetchOptions: { withCredentials: true },
  file: null,
  initialState: defaultState,
  onDataError: null,
  onDataLoaded: null,
  onPause: null,
  onReset: null,
  onResume: null,
  scrollRequestPosition: 'near',
  scrollTimeStamp: null,
};

export default NetworkProvider;
