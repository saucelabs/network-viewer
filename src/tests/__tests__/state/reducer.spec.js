import { Map, List } from 'immutable';

import * as types from './../../../state/network/types';
import { reducer, initialState } from './../../../state/network/reducer';
import networkDataMock from './../../__fixtures__/network.json';
import preparedMockData from './../../__fixtures__/preparedData';

describe('network reducer', () => {
  let state;

  it('UPDATE_DATA', () => {
    state = reducer(initialState, {
      type: types.UPDATE_DATA,
      payload: networkDataMock.log.entries,
    });
    expect(state.get('data').toJS()).toMatchSnapshot();
    expect(state.get('actualData').size).toMatchSnapshot();
  });

  it('UPDATE_SEARCH', () => {
    const newState = initialState.merge(new Map({
      actualData: preparedMockData,
    }));

    state = reducer(newState, {
      type: types.UPDATE_SEARCH,
      payload: {
        key: 'url',
        value: 'e96c15f68c68',
      },
    });
    expect(state.get('data').toJS()).toMatchSnapshot();
    expect(state.get('search')).toMatchSnapshot();
  });

  it('UPDATE_FILTER', () => {
    const newState = initialState.merge(new Map({
      actualData: preparedMockData,
    }));

    state = reducer(newState, {
      type: types.UPDATE_FILTER,
      payload: {
        key: 'type',
        value: 'html',
      },
    });
    expect(state.get('data').toJS()).toMatchSnapshot();
    expect(state.get('filter')).toMatchSnapshot();
  });

  it('UPDATE_SORT', () => {
    const newState = initialState.merge(new Map({
      data: preparedMockData,
    }));

    state = reducer(newState, {
      type: types.UPDATE_SORT,
      payload: {
        key: 'size',
        isAcs: true,
      },
    });
    expect(state.get('data').toJS()).toMatchSnapshot();
    expect(state.get('sort')).toMatchSnapshot();
  });

  it('ADD_NOTIFICATION', () => {
    const newState = initialState;

    state = reducer(newState, {
      type: types.ADD_NOTIFICATION,
      payload: {
        type: 'danger',
        title: 'Error notification',
      },
    });

    expect(state.get('notifications').toJS()).toMatchSnapshot();
  });

  it('DISMISS_NOTIFICATION', () => {
    const newState = initialState.merge(new Map({
      notifications: new List([{
        id: 1,
        type: 'danger',
      }]),
    }));

    state = reducer(newState, {
      type: types.DISMISS_NOTIFICATION,
      payload: 1,
    });

    expect(state.get('notifications').toJS()).toMatchSnapshot();
  });
});
