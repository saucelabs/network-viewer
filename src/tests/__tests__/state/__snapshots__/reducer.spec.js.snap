// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`network reducer ADD_NOTIFICATION 1`] = `
Array [
  Object {
    "id": 1,
    "title": "Error notification",
    "type": "danger",
  },
]
`;

exports[`network reducer DISMISS_NOTIFICATION 1`] = `Array []`;

exports[`network reducer UPDATE_DATA 1`] = `
Array [
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "slice",
    "index": 0,
    "method": "GET",
    "size": 135.04,
    "startedDateTime": "2019-10-14T08:47:11.592Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 4.382999992230907,
      "blocked": 5.496999992230907,
      "connect": -1,
      "dns": -1,
      "receive": 3.252000024076551,
      "send": 0.2819999999999998,
      "ssl": -1,
      "startTime": 0,
      "wait": 26.03899999633059,
    },
    "type": "document",
    "url": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
    "index": 1,
    "method": "GET",
    "size": 33.02,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 12.528999999631196,
      "blocked": 13.952999999631196,
      "connect": -1,
      "dns": -1,
      "receive": 7.124000025214627,
      "send": 0.2370000000000001,
      "ssl": -1,
      "startTime": 49,
      "wait": 18.80099998585507,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    "index": 2,
    "method": "GET",
    "size": 33.27,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 13.034000003244728,
      "blocked": 14.968000003244727,
      "connect": -1,
      "dns": -1,
      "receive": 1.9629999878816307,
      "send": 0.40400000000000014,
      "ssl": -1,
      "startTime": 49,
      "wait": 35.81200000759028,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  },
  Object {
    "body": ".logo{display:block;grid-area:I;width:200px;height:44px;margin-right:24px}",
    "domain": "developer.mozilla.org",
    "filename": "react-header.2a659f6ad280.css",
    "index": 3,
    "method": "GET",
    "size": 2.76,
    "startedDateTime": "2019-10-14T08:47:11.643Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.349999986123294,
      "blocked": 13.026999986123293,
      "connect": -1,
      "dns": -1,
      "receive": 1.951000012923032,
      "send": 0.21299999999999986,
      "ssl": -1,
      "startTime": 51,
      "wait": 36.02600000113249,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-header.2a659f6ad280.css",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "react-mdn.c66d8d89d98e.css",
    "index": 4,
    "method": "GET",
    "size": 142.88,
    "startedDateTime": "2019-10-14T08:47:11.644Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.250000010477379,
      "blocked": 13.423000010477379,
      "connect": -1,
      "dns": -1,
      "receive": 1.2719999940600246,
      "send": 0.5390000000000001,
      "ssl": -1,
      "startTime": 52,
      "wait": 37.96100001042895,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-mdn.c66d8d89d98e.css",
  },
]
`;

exports[`network reducer UPDATE_DATA 2`] = `5`;

exports[`network reducer UPDATE_FILTER 1`] = `
Array [
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "slice",
    "index": 0,
    "method": "GET",
    "size": 135.04,
    "startedDateTime": "2019-10-14T08:47:11.592Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 4.382999992230907,
      "blocked": 5.496999992230907,
      "connect": -1,
      "dns": -1,
      "receive": 3.252000024076551,
      "send": 0.2819999999999998,
      "ssl": -1,
      "startTime": 0,
      "wait": 26.03899999633059,
    },
    "type": "document",
    "url": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
    "index": 1,
    "method": "GET",
    "size": 33.02,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 12.528999999631196,
      "blocked": 13.952999999631196,
      "connect": -1,
      "dns": -1,
      "receive": 7.124000025214627,
      "send": 0.2370000000000001,
      "ssl": -1,
      "startTime": 49,
      "wait": 18.80099998585507,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    "index": 2,
    "method": "GET",
    "size": 33.27,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 13.034000003244728,
      "blocked": 14.968000003244727,
      "connect": -1,
      "dns": -1,
      "receive": 1.9629999878816307,
      "send": 0.40400000000000014,
      "ssl": -1,
      "startTime": 49,
      "wait": 35.81200000759028,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  },
  Object {
    "body": ".logo{display:block;grid-area:I;width:200px;height:44px;margin-right:24px}",
    "domain": "developer.mozilla.org",
    "filename": "react-header.2a659f6ad280.css",
    "index": 3,
    "method": "GET",
    "size": 2.76,
    "startedDateTime": "2019-10-14T08:47:11.643Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.349999986123294,
      "blocked": 13.026999986123293,
      "connect": -1,
      "dns": -1,
      "receive": 1.951000012923032,
      "send": 0.21299999999999986,
      "ssl": -1,
      "startTime": 51,
      "wait": 36.02600000113249,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-header.2a659f6ad280.css",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "react-mdn.c66d8d89d98e.css",
    "index": 4,
    "method": "GET",
    "size": 142.88,
    "startedDateTime": "2019-10-14T08:47:11.644Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.250000010477379,
      "blocked": 13.423000010477379,
      "connect": -1,
      "dns": -1,
      "receive": 1.2719999940600246,
      "send": 0.5390000000000001,
      "ssl": -1,
      "startTime": 52,
      "wait": 37.96100001042895,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-mdn.c66d8d89d98e.css",
  },
]
`;

exports[`network reducer UPDATE_FILTER 2`] = `
Object {
  "key": "type",
  "value": "html",
}
`;

exports[`network reducer UPDATE_SEARCH 1`] = `Array []`;

exports[`network reducer UPDATE_SEARCH 2`] = `
Object {
  "key": "url",
  "value": "e96c15f68c68",
}
`;

exports[`network reducer UPDATE_SORT 1`] = `
Array [
  Object {
    "body": ".logo{display:block;grid-area:I;width:200px;height:44px;margin-right:24px}",
    "domain": "developer.mozilla.org",
    "filename": "react-header.2a659f6ad280.css",
    "index": 3,
    "method": "GET",
    "size": 2.76,
    "startedDateTime": "2019-10-14T08:47:11.643Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.349999986123294,
      "blocked": 13.026999986123293,
      "connect": -1,
      "dns": -1,
      "receive": 1.951000012923032,
      "send": 0.21299999999999986,
      "ssl": -1,
      "startTime": 51,
      "wait": 36.02600000113249,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-header.2a659f6ad280.css",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
    "index": 1,
    "method": "GET",
    "size": 33.02,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 12.528999999631196,
      "blocked": 13.952999999631196,
      "connect": -1,
      "dns": -1,
      "receive": 7.124000025214627,
      "send": 0.2370000000000001,
      "ssl": -1,
      "startTime": 49,
      "wait": 18.80099998585507,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    "index": 2,
    "method": "GET",
    "size": 33.27,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 13.034000003244728,
      "blocked": 14.968000003244727,
      "connect": -1,
      "dns": -1,
      "receive": 1.9629999878816307,
      "send": 0.40400000000000014,
      "ssl": -1,
      "startTime": 49,
      "wait": 35.81200000759028,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "slice",
    "index": 0,
    "method": "GET",
    "size": 135.04,
    "startedDateTime": "2019-10-14T08:47:11.592Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 4.382999992230907,
      "blocked": 5.496999992230907,
      "connect": -1,
      "dns": -1,
      "receive": 3.252000024076551,
      "send": 0.2819999999999998,
      "ssl": -1,
      "startTime": 0,
      "wait": 26.03899999633059,
    },
    "type": "document",
    "url": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "react-mdn.c66d8d89d98e.css",
    "index": 4,
    "method": "GET",
    "size": 142.88,
    "startedDateTime": "2019-10-14T08:47:11.644Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.250000010477379,
      "blocked": 13.423000010477379,
      "connect": -1,
      "dns": -1,
      "receive": 1.2719999940600246,
      "send": 0.5390000000000001,
      "ssl": -1,
      "startTime": 52,
      "wait": 37.96100001042895,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-mdn.c66d8d89d98e.css",
  },
]
`;

exports[`network reducer UPDATE_SORT 2`] = `
Object {
  "isAcs": true,
  "key": "size",
}
`;
